static/src/js/components/Map/GranuleGridLayer.js

```diff
@@ -19,107 +15,292 @@ import './GranuleGridLayer.scss'
+const updateGranuleGridLayer = (instance, props, prevProps) => {
+  const layers = instance._layers // List of layers
+
+  const {
+    drawingNewLayer,
+    focusedCollectionId,
+    focusedGranuleId,
+    projection,
+    project,
+    imageryCache,
+    isProjectPage,
+    onChangeFocusedGranule,
+    onExcludeGranule,
+    onMetricsMap
+  } = props
+
+  const {
+    drawingNewLayer: prevDrawingNewLayer,
+    focusedGranuleId: prevFocusedGranule,
+    focusedCollectionId: prevFocusedCollection,
+    isProjectPage: prevIsProjectPage
+  } = prevProps
+
+  const prevLayerData = getLayerData(prevProps)
+  const layerData = getLayerData(props)
+
+  const layerDataCollectionIds = Object.keys(layerData)
+
+  // Nothing should be drawn, remove any existing layers
+  if (layerDataCollectionIds.length === 0) {
+    Object.values(layers).forEach((layer) => instance.removeLayer(layer))
+  } else if (layerDataCollectionIds.length < Object.keys(prevLayerData).length) {
+    // If there is less data than before, figure out which collection was removed and remove the layer
+
+    const prevIds = Object.keys(prevLayerData)
+    const diffIds = difference(prevIds, layerDataCollectionIds)
+
+    diffIds.forEach((collectionId) => {
+      Object.values(layers).forEach((layer) => {
+        if (layer.collectionId === collectionId) {
+          instance.removeLayer(layer)
+        }
+      })
+    })
+  }
+
+  // Loop through each layer data object to update the layer
+  layerDataCollectionIds.forEach((id) => {
+    const {
+      collectionId,
+      color,
+      lightColor,
+      metadata,
+      isVisible,
+      granules = {}
+    } = layerData[id]
+
+    // Find the layer for this collection
+    const [layer] = Object.values(layers).filter((l) => l.collectionId === collectionId)
+
+    const { project: prevPropsProject } = prevProps
+    const { collections: prevPropsProjectCollections = {} } = prevPropsProject
+    const { byId: prevPropsProjectCollectionsById = {} } = prevPropsProjectCollections
+    const { [collectionId]: prevProjectCollection = {} } = prevPropsProjectCollectionsById
+    const { granules: prevProjectGranules = {} } = prevProjectCollection
+
+    const { collections: projectCollections = {} } = project
+    const { byId: projectCollectionsById = {} } = projectCollections
+    const { [collectionId]: projectCollection = {} } = projectCollectionsById
+    const { granules: newGranules = {} } = projectCollection
+
+    let prevAddedGranuleIds
+    let prevRemovedGranuleIds
+    let addedGranuleIds
+    let removedGranuleIds
+
+    if (prevProjectGranules) {
+      ({
+        addedGranuleIds: prevAddedGranuleIds,
+        removedGranuleIds: prevRemovedGranuleIds
+      } = prevProjectGranules)
+    }
+
+    if (newGranules) {
+      ({
+        addedGranuleIds,
+        removedGranuleIds
+      } = newGranules)
+    }
+
+    // If there are no granules, bail out
+    const { byId: granulesById = {} } = granules
```

Comment:
I might be reading this wrong.  Do you want to do an early return if granulesById is empty?